### PR TITLE
feat: pagination

### LINKED ISSUE #1382: Add pagination to the Request Node /getTransactionsByChannelId endpoint

### PR DESCRIPTION
**Pull request by rodrigopavezi**
Fixes #1382

Add pagination all the way from the request-client to the request-node and subgraph queries. 

## Release Notes

- **New Features**
	- Enhanced pagination capabilities for retrieving channels and requests by topic, allowing users to specify the number of results per page and the page number.
	- Updated multiple methods across the application to support pagination, improving data retrieval efficiency.

- **Bug Fixes**
	- Retained existing error handling and control flow, ensuring stability in the updated methods.

- **Documentation**
	- Updated method signatures and documentation to reflect the new pagination parameters for clarity and usability.

### CODE REVIEW

File: packages/data-access/src/data-read.ts [hunk 1]
```diff
@@ -49,15 +49,19 @@ export class DataAccessRead implements DataAccessTypes.IDataRead {
...
   }
 
   async getChannelsByMultipleTopics(
     topics: string[],
     updatedBetween?: DataAccessTypes.ITimestampBoundaries,
+    page?: number,
+    pageSize?: number,
   ): Promise<DataAccessTypes.IReturnGetChannelsByTopic> {
-    const result = await this.storage.getTransactionsByTopics(topics);
+    const result = await this.storage.getTransactionsByTopics(topics, page, pageSize);
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/data-read.ts:64]: Could you adjust the changes so the page validation will enforce page to be greater or equal 1?

File: packages/request-client.js/src/http-data-access.ts [hunk 2]
```diff
@@ -191,10 +195,14 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
     topics: string[],
     updatedBetween?: DataAccessTypes.ITimestampBoundaries,
+    page?: number,
+    pageSize?: number,
   ): Promise<DataAccessTypes.IReturnGetChannelsByTopic> {
     return await this.fetchAndRetry('/getChannelsByMultipleTopics', {
       topics,
       updatedBetween,
+      page,
+      pageSize,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/http-data-access.ts:205]: I will fix this

File: packages/request-client.js/src/http-data-access.ts [hunk 3]
```diff
@@ -175,10 +175,14 @@
...
     topic: string,
     updatedBetween?: DataAccessTypes.ITimestampBoundaries,
+    page?: number,
+    pageSize?: number,
   ): Promise<DataAccessTypes.IReturnGetChannelsByTopic> {
     return await this.fetchAndRetry('/getChannelsByTopic', {
       topic,
       updatedBetween,
+      page,
+      pageSize,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/http-data-access.ts:185]: I will fix this

File: packages/transaction-manager/test/index.test.ts [hunk 4]
```diff
@@ -1491,6 +1521,8 @@
       expect(fakeDataAccess.getChannelsByMultipleTopics).toHaveBeenCalledWith(
         [extraTopics[0]],
         undefined,
+        undefined,
+        undefined,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/transaction-manager/test/index.test.ts:1525]: I will change that

File: packages/thegraph-data-access/src/subgraph-client.ts [hunk 5]
```diff
@@ -50,10 +50,21 @@ export class SubgraphClient implements StorageTypes.IIndexer {
...
+    if (pageSize !== undefined && pageSize <= 0) {
+      throw new Error('Page size must be greater than 0');
+    }
+
+    const skip = page && pageSize ? (page - 1) * pageSize : null;
+    const first = pageSize || null;
     const { _meta, channels } = await this.graphql.request<
       Meta & { channels: { transactions: Transaction[] }[] }
-    >(GetTransactionsByTopics, { topics });
+    >(GetTransactionsByTopics, { topics, first, skip });
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/thegraph-data-access/src/subgraph-client.ts:67]: Good point, I will fix that

File: packages/data-access/src/in-memory-indexer.ts [hunk 6]
```diff
@@ -55,8 +55,33 @@ export class InMemoryIndexer implements StorageTypes.IIndexer {
...
+    if (page && pageSize) {
+      const start = (page - 1) * pageSize;
+      // Return empty result if page exceeds available data
+      if (start >= total) {
+        return {
+          blockNumber: 0,
+          transactions: [],
+          pagination: { total, page, pageSize },
+        };
+      }
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/in-memory-indexer.ts:82]: Good point, I will fix that

File: packages/data-access/src/in-memory-indexer.ts [hunk 7]
```diff
@@ -55,8 +55,33 @@ export class InMemoryIndexer implements StorageTypes.IIndexer {
...
+      channelIds = channelIds.slice(start, start + pageSize);
+    }
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/in-memory-indexer.ts:84]: good one, I will fix that

File: packages/transaction-manager/test/index.test.ts [hunk 8]
```diff
@@ -1466,7 +1491,28 @@ describe('index', () => {
...
+      const page1 = await transactionManager.getChannelsByTopic(extraTopics[0], undefined, 1, 2);
+      expect(page1.result.transactions).toHaveLength(2);
+
+      // Test second page
+      const page2 = await transactionManager.getChannelsByTopic(extraTopics[0], undefined, 2, 2);
+      expect(page2.result.transactions).toHaveLength(2);
+
+      // Test last page
+      const lastPage = await transactionManager.getChannelsByTopic(extraTopics[0], undefined, 3, 2);
+      expect(lastPage.result.transactions).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/transaction-manager/test/index.test.ts:1515]: I will fix this

File: packages/data-access/src/data-read.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/data-read.ts:64]: Could you now please create a commit change for this adjustment?

[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/data-read.ts:64]: I am not sure about the pendingItems. Should I replace it with the your suggested one?

File: packages/thegraph-data-access/src/subgraph-client.ts [hunk 9]
```diff
@@ -50,10 +54,30 @@ export class SubgraphClient implements StorageTypes.IIndexer {
...
+    const skip = (effectivePage - 1) * effectivePageSize;
+
     const { _meta, channels } = await this.graphql.request<
       Meta & { channels: { transactions: Transaction[] }[] }
-    >(GetTransactionsByTopics, { topics });
+    >(GetTransactionsByTopics, {
+      topics,
+      first: effectivePageSize,
+      skip,
+    });
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/thegraph-data-access/src/subgraph-client.ts:80]: I will fix that

File: packages/thegraph-data-access/src/subgraph-client.ts [hunk 10]
```diff
@@ -50,19 +54,46 @@ export class SubgraphClient implements StorageTypes.IIndexer {
...
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/thegraph-data-access/src/subgraph-client.ts:80]: I will ignore that for now

File: packages/types/src/data-access-types.ts [hunk 11]
```diff
@@ -13,10 +13,14 @@ export interface IDataRead {
   getChannelsByTopic: (
     topic: string,
     updatedBetween?: ITimestampBoundaries,
+    page?: number,
+    pageSize?: number,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/types/src/data-access-types.ts:17]: We don't need that now

File: packages/types/src/data-access-types.ts [hunk 12]
```diff
@@ -69,6 +73,13 @@ export type PersistTransactionEmitter = ConfirmationEventEmitter<IReturnPersistT
 
 export type IReturnPersistTransaction = PersistTransactionEmitter & IReturnPersistTransactionRaw;
 
+interface PaginationMetadata {
+  total: number; // Total number of items available
+  page?: number; // Current page number if pagination was used
+  pageSize?: number; // Page size if pagination was used
+  hasMore: boolean; // Whether there are more items available
+}
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/types/src/data-access-types.ts]: I will fix that

File: packages/data-access/src/data-read.ts [hunk 13]
```diff
@@ -73,48 +85,87 @@ export class DataAccessRead implements DataAccessTypes.IDataRead {
...
-        ),
+        storageMeta,
+        transactionsStorageLocation,
+        pagination:
+          page && pageSize
+            ? {
+                total: result.transactions.length,
+                page,
+                pageSize,
+                hasMore: page * pageSize < result.transactions.length + pendingItemsOnCurrentPage,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/data-access/src/data-read.ts]: I will fix that

File: packages/request-client.js/src/http-data-access.ts [hunk 14]
```diff
@@ -175,11 +175,34 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
+    page?: number,
+    pageSize?: number,
   ): Promise<DataAccessTypes.IReturnGetChannelsByTopic> {
-    return await this.fetchAndRetry('/getChannelsByTopic', {
+    if (page !== undefined && page < 1) {
+      throw new Error(`Page number must be greater than or equal to 1 but it is ${page}`);
+    }
+    if (pageSize !== undefined && pageSize <= 0) {
+      throw new Error(`Page size must be positive but it is ${pageSize}`);
+    }
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/http-data-access.ts:186]: I will fix that

File: packages/request-client.js/src/http-data-access.ts [hunk 15]
```diff
@@ -175,11 +175,34 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
+    };
+
+    if (page !== undefined) {
+      params.page = page;
+      if (pageSize !== undefined) {
+        params.pageSize = pageSize;
+      }
+    }
+
+    return await this.fetchAndRetry('/getChannelsByTopic', params);
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/http-data-access.ts:205]: I will fix that

File: packages/request-client.js/src/api/request-network.ts [hunk 16]
```diff
@@ -361,11 +381,21 @@ export default class RequestNetwork {
...
   ): Promise<Request[]> {
     // Gets all the requests indexed by the value of the identity
     const requestsAndMeta: RequestLogicTypes.IReturnGetRequestsByTopic =
-      await this.requestLogic.getRequestsByMultipleTopics(topics, updatedBetween);
+      await this.requestLogic.getRequestsByMultipleTopics(
+        topics,
+        updatedBetween,
+        options?.page,
+        options?.pageSize,
+      );
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/api/request-network.ts:398]: I will change that

File: packages/request-client.js/src/api/request-network.ts [hunk 17]
```diff
@@ -309,11 +319,21 @@ export default class RequestNetwork {
...
   ): Promise<Request[]> {
     // Gets all the requests indexed by the value of the identity
     const requestsAndMeta: RequestLogicTypes.IReturnGetRequestsByTopic =
-      await this.requestLogic.getRequestsByTopic(topic, updatedBetween);
+      await this.requestLogic.getRequestsByTopic(
+        topic,
+        updatedBetween,
+        options?.page,
+        options?.pageSize,
+      );
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-client.js/src/api/request-network.ts:336]: I will change that

File: packages/transaction-manager/test/index.test.ts [hunk 18]
```diff
@@ -1491,6 +1542,40 @@ describe('index', () => {
...
+        1, // page
+        2, // pageSize
+      );
+
+      expect(Object.keys(result.result.transactions)).toHaveLength(2);
+      expect(fakeDataAccess.getChannelsByMultipleTopics).toHaveBeenCalledWith(
+        [extraTopics[0], extraTopics[1]],
+        undefined,
+        1,
+        2,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/transaction-manager/test/index.test.ts:1578]: I will leave it for another time

File: packages/request-node/test/getChannelsByTopic.test.ts [hunk 19]
```diff
@@ -98,6 +98,19 @@ describe('getChannelsByTopic', () => {
...
+      .get('/getChannelsByTopic')
+      .query({ topic: commonTopic, page: 1, pageSize: 1 })
+      .set('Accept', 'application/json')
+      .expect(StatusCodes.OK);
+
+    expect(serverResponse.body.result.transactions).toMatchObject(
+      expect.objectContaining({
+        [channelId]: [expect.objectContaining({ transaction: transactionData })],
+      }),
+    );
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-node/test/getChannelsByTopic.test.ts:113]: I will fix that

**Review by aimensahnoun — APPROVED** [review]

**Pull request merged**